File: mobile/android/app/src/main/java/com/trickle/faucet/viewmodel/FaucetViewModel.kt
package com.trickle.faucet.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.walletconnect.android.Core
import com.walletconnect.sign.client.Sign
import com.walletconnect.sign.client.SignClient
import com.walletconnect.web3.modal.client.Web3Modal
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import org.web3j.crypto.Credentials
import org.web3j.protocol.Web3j
import org.web3j.protocol.http.HttpService
import org.web3j.tx.gas.DefaultGasProvider
import java.math.BigInteger

data class FaucetUiState(
    val isConnected: Boolean = false,
    val walletAddress: String? = null,
    val isLoading: Boolean = false,
    val errorMessage: String? = null,
    val successMessage: String? = null,
    val totalClaims: Int = 0,
    val dailyClaims: Int = 0,
    val canClaim: Boolean = true,
    val nextClaimTime: String? = null
)

class FaucetViewModel : ViewModel() {

    private val _uiState = MutableStateFlow(FaucetUiState())
    val uiState: StateFlow<FaucetUiState> = _uiState.asStateFlow()

    private val web3j = Web3j.build(HttpService("https://mainnet.base.org"))
    private val faucetContractAddress = "0x8D08e77837c28fB271D843d84900544cA46bA2F3"
    
    private var currentSession: Sign.Model.Session? = null

    init {
        setupWalletConnectCallbacks()
        loadStats()
    }

    private fun setupWalletConnectCallbacks() {
        val wcDelegate = object : SignClient.WalletDelegate {
            override fun onSessionApproved(approvedSession: Sign.Model.Session) {
                currentSession = approvedSession
                val address = approvedSession.accounts.firstOrNull()?.split(":")?.lastOrNull()
                _uiState.value = _uiState.value.copy(
                    isConnected = true,
                    walletAddress = address,
                    isLoading = false,
                    successMessage = "Wallet connected successfully!"
                )
                clearMessages()
            }

            override fun onSessionRejected(rejectedSession: Sign.Model.SessionRequestResponse) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    errorMessage = "Connection rejected by wallet"
                )
                clearMessages()
            }

            override fun onSessionDelete(deletedSession: Sign.Model.SessionDelete) {
                currentSession = null
                _uiState.value = _uiState.value.copy(
                    isConnected = false,
                    walletAddress = null,
                    successMessage = "Wallet disconnected"
                )
                clearMessages()
            }

            override fun onSessionExtend(session: Sign.Model.Session) {
                currentSession = session
            }

            override fun onSessionRequestResponse(response: Sign.Model.SessionRequestResponse) {
                // Handle transaction responses
                when (response) {
                    is Sign.Model.SessionRequestResponse.Result -> {
                        _uiState.value = _uiState.value.copy(
                            isLoading = false,
                            successMessage = "Transaction submitted successfully!"
                        )
                        loadStats() // Refresh stats after successful transaction
                    }
                    is Sign.Model.SessionRequestResponse.Error -> {
                        _uiState.value = _uiState.value.copy(
                            isLoading = false,
                            errorMessage = "Transaction failed: ${response.error.message}"
                        )
                    }
                }
                clearMessages()
            }

            override fun onConnectionStateChange(state: Sign.Model.ConnectionState) {
                // Handle connection state changes
            }

            override fun onError(error: Sign.Model.Error) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    errorMessage = "WalletConnect error: ${error.throwable.message}"
                )
                clearMessages()
            }
        }

        SignClient.setWalletDelegate(wcDelegate)
    }

    fun connectWallet() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)
            
            try {
                // Open Web3Modal
                Web3Modal.open { error ->
                    if (error != null) {
                        _uiState.value = _uiState.value.copy(
                            isLoading = false,
                            errorMessage = "Failed to open wallet selection: ${error.throwable.message}"
                        )
                        clearMessages()
                    }
                }
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    errorMessage = "Connection failed: ${e.message}"
                )
                clearMessages()
            }
        }
    }

    fun disconnectWallet() {
        viewModelScope.launch {
            currentSession?.let { session ->
                val disconnect = Sign.Params.Disconnect(
                    sessionTopic = session.topic,
                    reason = Sign.Model.DeletedSession.Reason(
                        code = 6000,
                        message = "User disconnected"
                    )
                )
                
                SignClient.disconnect(disconnect) { error ->
                    error?.let {
                        _uiState.value = _uiState.value.copy(
                            errorMessage = "Disconnect failed: ${it.throwable.message}"
                        )
                        clearMessages()
                    }
                }
            }
        }
    }

    fun claimTokens() {
        viewModelScope.launch {
            val address = _uiState.value.walletAddress
            if (address == null || currentSession == null) {
                _uiState.value = _uiState.value.copy(
                    errorMessage = "Please connect your wallet first"
                )
                clearMessages()
                return@launch
            }

            _uiState.value = _uiState.value.copy(isLoading = true)

            try {
                // Check eligibility first
                val eligibility = checkEligibility(address)
                if (!eligibility.first) {
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        errorMessage = eligibility.second,
                        canClaim = false,
                        nextClaimTime = eligibility.third
                    )
                    clearMessages()
                    return@launch
                }

                // Create transaction request
                val transaction = createClaimTransaction(address)
                
                val sessionRequest = Sign.Params.SessionRequest(
                    sessionTopic = currentSession!!.topic,
                    request = Sign.Model.SessionRequest(
                        id = System.currentTimeMillis(),
                        method = "eth_sendTransaction",
                        params = listOf(transaction)
                    ),
                    chainId = "eip155:8453" // Base chain
                )

                SignClient.request(sessionRequest) { error ->
                    if (error != null) {
                        _uiState.value = _uiState.value.copy(
                            isLoading = false,
                            errorMessage = "Transaction request failed: ${error.throwable.message}"
                        )
                        clearMessages()
                    }
                }

            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    errorMessage = "Claim failed: ${e.message}"
                )
                clearMessages()
            }
        }
    }

    private suspend fun checkEligibility(address: String): Triple<Boolean, String, String?> {
        // Implementation to check eligibility via API
        // This should call your backend API
        return try {
            // Mock implementation - replace with actual API call
            Triple(true, "", null)
        } catch (e: Exception) {
            Triple(false, "Eligibility check failed: ${e.message}", null)
        }
    }

    private fun createClaimTransaction(address: String): Map<String, String> {
        // Create transaction data for requestTokens() function call
        val functionSignature = "0x9fbc8713" // keccak256("requestTokens()")
        
        return mapOf(
            "from" to address,
            "to" to faucetContractAddress,
            "data" to functionSignature,
            "value" to "0x0"
        )
    }

    private fun loadStats() {
        viewModelScope.launch {
            try {
                // Mock implementation - replace with actual API call to your backend
                _uiState.value = _uiState.value.copy(
                    totalClaims = 0,
                    dailyClaims = 0
                )
            } catch (e: Exception) {
                // Handle error silently for stats
            }
        }
    }

    fun handleAppResume() {
        // Handle any pending wallet responses when app resumes
        // This is useful for deep linking scenarios
    }

    private fun clearMessages() {
        viewModelScope.launch {
            kotlinx.coroutines.delay(3000) // Show message for 3 seconds
            _uiState.value = _uiState.value.copy(
                errorMessage = null,
                successMessage = null
            )
        }
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(errorMessage = null)
    }

    fun clearSuccess() {
        _uiState.value = _uiState.value.copy(successMessage = null)
    }
}

File: mobile/ios/TrickleFaucet/FaucetViewModel.swift
import SwiftUI
import Foundation

@MainActor
class FaucetViewModel: ObservableObject {
    @Published var isConnected = false
    @Published var walletAddress: String?
    @Published var isLoading = false
    @Published var errorMessage: String?
    @Published var successMessage: String?
    @Published var totalClaims = 0
    @Published var dailyClaims = 0
    @Published var canClaim = true
    @Published var nextClaimTime: String?
    
    private let faucetContractAddress = "0x8D08e77837c28fB271D843d84900544cA46bA2F3"
    private let baseRpcUrl = "https://mainnet.base.org"
    
    func connectWallet(using walletConnectManager: WalletConnectManager) {
        isLoading = true
        clearMessages()
        
        Task {
            do {
                let session = try await walletConnectManager.connect()
                if let account = session.accounts.first {
                    // Extract address from account (format: eip155:8453:0x...)
                    let address = String(account.split(separator: ":").last ?? "")
                    await MainActor.run {
                        self.walletAddress = address
                        self.isConnected = true
                        self.isLoading = false
                        self.successMessage = "Wallet connected successfully!"
                        self.clearMessagesAfterDelay()
                    }
                }
            } catch {
                await MainActor.run {
                    self.isLoading = false
                    self.errorMessage = "Connection failed: \(error.localizedDescription)"
                    self.clearMessagesAfterDelay()
                }
            }
        }
    }
    
    func disconnectWallet(using walletConnectManager: WalletConnectManager) {
        Task {
            do {
                try await walletConnectManager.disconnect()
                await MainActor.run {
                    self.isConnected = false
                    self.walletAddress = nil
                    self.successMessage = "Wallet disconnected"
                    self.clearMessagesAfterDelay()
                }
            } catch {
                await MainActor.run {
                    self.errorMessage = "Disconnect failed: \(error.localizedDescription)"
                    self.clearMessagesAfterDelay()
                }
            }
        }
    }
    
    func claimTokens(using walletConnectManager: WalletConnectManager) {
        guard let address = walletAddress else {
            errorMessage = "Please connect your wallet first"
            clearMessagesAfterDelay()
            return
        }
        
        isLoading = true
        clearMessages()
        
        Task {
            do {
                // Check eligibility first
                let eligibility = try await checkEligibility(address: address)
                if !eligibility.eligible {
                    await MainActor.run {
                        self.isLoading = false
                        self.errorMessage = eligibility.message
                        self.canClaim = false
                        self.nextClaimTime = eligibility.nextClaimTime
                        self.clearMessagesAfterDelay()
                    }
                    return
                }
                
                // Create transaction
                let transaction = createClaimTransaction(from: address)
                
                // Send transaction via WalletConnect
                let txHash = try await walletConnectManager.sendTransaction(transaction)
                
                // Log the claim
                try await logClaim(address: address, txHash: txHash)
                
                await MainActor.run {
                    self.isLoading = false
                    self.successMessage = "Claim successful! Transaction: \(txHash.prefix(10))..."
                    self.loadStats()
                    self.clearMessagesAfterDelay()
                }
                
            } catch {
                await MainActor.run {
                    self.isLoading = false
                    self.errorMessage = "Claim failed: \(error.localizedDescription)"
                    self.clearMessagesAfterDelay()
                }
            }
        }
    }
    
    func loadStats() {
        Task {
            do {
                let stats = try await fetchStats()
                await MainActor.run {
                    self.totalClaims = stats.totalClaims
                    self.dailyClaims = stats.dailyClaims
                }
            } catch {
                // Handle error silently for stats
                print("Failed to load stats: \(error)")
            }
        }
    }
    
    // MARK: - Private Methods
    
    private func checkEligibility(address: String) async throws -> (eligible: Bool, message: String, nextClaimTime: String?) {
        // Replace with your actual API endpoint
        guard let url = URL(string: "https://your-api-endpoint.com/check-eligibility") else {
            throw URLError(.badURL)
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body = ["address": address]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        
        let (data, _) = try await URLSession.shared.data(for: request)
        
        struct EligibilityResponse: Codable {
            let eligible: Bool
            let message: String?
            let nextClaimTime: String?
        }
        
        let response = try JSONDecoder().decode(EligibilityResponse.self, from: data)
        return (response.eligible, response.message ?? "", response.nextClaimTime)
    }
    
    private func createClaimTransaction(from address: String) -> [String: Any] {
        // Function signature for requestTokens()
        let functionSignature = "0x9fbc8713"
        
        return [
            "from": address,
            "to": faucetContractAddress,
            "data": functionSignature,
            "value": "0x0"
        ]
    }
    
    private func logClaim(address: String, txHash: String) async throws {
        // Replace with your actual API endpoint
        guard let url = URL(string: "https://your-api-endpoint.com/log-claim") else {
            throw URLError(.badURL)
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body = ["address": address, "txHash": txHash]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        
        let (_, _) = try await URLSession.shared.data(for: request)
    }
    
    private func fetchStats() async throws -> (totalClaims: Int, dailyClaims: Int) {
        // Replace with your actual API endpoint
        guard let url = URL(string: "https://your-api-endpoint.com/stats") else {
            throw URLError(.badURL)
        }
        
        let (data, _) = try await URLSession.shared.data(from: url)
        
        struct StatsResponse: Codable {
            let totalClaims: Int
            let claimsLast24h: Int
        }
        
        let response = try JSONDecoder().decode(StatsResponse.self, from: data)
        return (response.totalClaims, response.claimsLast24h)
    }
    
    private func clearMessages() {
        errorMessage = nil
        successMessage = nil
    }
    
    private func clearMessagesAfterDelay() {
        Task {
            try await Task.sleep(nanoseconds: 3_000_000_000) // 3 seconds
            await MainActor.run {
                self.clearMessages()
            }
        }
    }
}